File: Python/s/Node.py
# coding=utf-8
from Function import Function,FunctionType
from Util import Util

class Node:
    """docstring for Node"""
    def __init__(self, first,rest):
        self.first = first
        self.rest = rest
        if rest==None:
            self.length=1
        else:
            self.length=rest.Length()+1

    def Length(self):
        return self.length

    def First(self):
        return self.first
    
    def Rest(self):
        return self.rest
    
    @staticmethod
    def toString_one(sb,o):
        if o==None:
            sb.append("[]")
        elif isinstance(o,Node):
            o.toString(sb)
        elif isinstance(o,basestring):
            sb.append(Util.stringToEscape(o,'"','"'))
        elif isinstance(o,Function):
            if o.Function_type()==FunctionType.Fun_buildIn:
                sb.append("'")
                sb.append(str(o))
            elif o.Function_type()==FunctionType.Fun_user:
                sb.append(str(o))
            elif o.Function_type()==FunctionType.Fun_cache:
                sb.append("[]")
            else:
                sb.append(str(o))
        elif isinstance(o,bool):
            if o:
                sb.append("true")
            else:
                sb.append("false")
        elif isinstance(o,int):
            sb.append(str(o))
        else:
            vx=str(o)
            if(vx==None):
                sb.append("[]")
            else:
                sb.append("'")
                sb.append(vx)
    
    def toString(self,sb):
        sb.append("[")
        Node.toString_one(sb,self.first)
        tmp=self.rest
        while tmp!=None:
            sb.append(" ")
            Node.toString_one(sb,tmp.First());
            tmp=tmp.Rest()
        sb.append("]")
        
    def __str__(self):
        sb=[];
        self.toString(sb)
        return "".join(sb)
    
    
    @staticmethod
    def extend(v,vs):
        return Node(v,vs)
    
    @staticmethod
    def kvs_extend(k,v,kvs):
        return Node(k,Node(v,kvs))
    
    @staticmethod
    def kvs_find1st(kvs,k):
        unfind=True
        r=None
        while(kvs!=None and unfind):
            key=kvs.First()
            kvs=kvs.Rest()
            if key==k:
                unfind=False
                r=kvs.First()
            kvs=kvs.Rest()
        return r
    
    @staticmethod
    def reverse(vs):
        r=None
        tmp=vs
        while tmp!=None:
            r=Node.extend(tmp.First(), r)
            tmp=tmp.Rest()
        return r
    
    @staticmethod
    def list(*args):
        r=None
        len_c=len(args)-1
        while len_c>-1:
            r=Node.extend(args[len_c],r)
            len_c=len_c-1
        return r

File: JAVA/s/src/main/java/st3/Node.java
package st3;

public class Node<T> {
    private Node(T v, Node<T> n) {
        this.first=v;
        this.rest=n;
        if(n!=null) {
            this.length=n.length+1;
        }else {
            this.length=1;
        }
    }
    public final int length;
    public final T first;
    public final Node<T> rest;

    public static <T> Node<T> extend(T v, Node<T> vs) {
        return new Node(v,vs);
    }
    public static <T> Node<T> reverse(Node<T> vs) {
        Node<T> rs=null;
        while (vs!=null){
            rs=extend(vs.first,rs);
            vs=vs.rest;
        }
        return rs;
    }
    public static Node<Object> list(Object ...ids) {
        Node<Object> r=null;
        for(int i=ids.length-1;i>-1;i--){
            r=Node.extend(ids[i],r);
        }
        return r;
    }
}
